### PR TITLE
[Flax] Add Electra models

### PR DESCRIPTION
**Pull request by CoderPat**
# What does this PR do?

- Implement Flax version of Electra model and classes for different downstream tasks: 
  - `FlaxElectraModel`
  - `FlaxElectraForMaskedLM`
  - `FlaxElectraForPreTraining`
  - `FlaxElectraForMultipleChoice`
  - `FlaxElectraForQuestionAnswering`
  - `FlaxElectraForSequenceClassification`
  - `FlaxElectraForTokenClassification`
 
Most of the code taken from FlaxBert code and the Pytorch Electra code, and also adapted code from the original PR from @chris-tng  (credit where it's due since he started this in #9172 ).

Running the tests (including the slow ones) works, and I already tested in my downstream task and seems to be working

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

Tagging @patrickvonplaten @sgugger @chris-tng , feel free to tag other people

### CODE REVIEW

File: src/transformers/models/electra/modeling_flax_electra.py [hunk 1]
```diff
@@ -0,0 +1,673 @@
...
+
+class FlaxElectraForMaskedLMModule(nn.Module):
+    config: ElectraConfig
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
+        self.electra = FlaxElectraModule(config=self.config, dtype=self.dtype, add_pooling_layer=False)
+        self.generator_predictions = FlaxElectraGeneratorPredictions(config=self.config)
+        # TODO: should we have the option to include shared embeddings here
+        self.generator_lm_head = nn.Dense(self.config.vocab_size, dtype=self.dtype)
```
Comment (CoderPat) [inline src/transformers/models/electra/modeling_flax_electra.py:485]: I didn't add the code for shared embeddings here since the pytorch version didn't have it either, should I still add it?
Also found it weird that the original pytorch code for bert doesn't seem to have it either or maybe I missed it

File: src/transformers/models/electra/modeling_flax_electra.py [hunk 2]
```diff
@@ -0,0 +1,673 @@
...
+    ELECTRA Model with a span classification head on top for extractive question-answering tasks like SQuAD (a linear
+    layers on top of the hidden-states output to compute `span start logits` and `span end logits`).
+    """,
+    ELECTRA_START_DOCSTRING,
+)
+class FlaxElectraForQuestionAnswering(FlaxElectraPreTrainedModel):
+    module_class = FlaxElectraForQuestionAnsweringModule
+
+
+class FlaxElectraForSequenceClassificationModule(nn.Module):
```
Comment (CoderPat) [inline src/transformers/models/electra/modeling_flax_electra.py:640]: Note that the structure of the ```FlaxElectraForSequenceClassificationModule``` differs abit from original PyTorch version, since I use a pooler class. However since the pretrained models don't include weights for this class it seems to be fine. Could this potentially be a problem if people want to convert from ```ElectraForSequenceClassification``` -> ```FlaxElectraForSequenceClassification```

File: src/transformers/modeling_flax_utils.py [hunk 3]
```diff
@@ -425,6 +432,133 @@ def save_pretrained(self, save_directory: Union[str, os.PathLike], push_to_hub=F
             logger.info(f"Model pushed to the hub in this commit: {url}")
 
 
+class SequenceSummary(nn.Module):
```
Comment (CoderPat) [inline src/transformers/modeling_flax_utils.py:435]: Didn't need to add this, but feel like it might be usefull when other classes get ported to Flax

File: src/transformers/modeling_flax_utils.py [hunk 4]
```diff
@@ -425,6 +432,133 @@ def save_pretrained(self, save_directory: Union[str, os.PathLike], push_to_hub=F
...
+
+        output = self.first_dropout(output, deterministic=deterministic)
+        output = self.summary(output)
+        output = self.activation(output)
+        output = self.last_dropout(output, deterministic=deterministic)
+
+        return output
+
+
+class TiedDense(nn.Module):
```
Comment (CoderPat) [inline src/transformers/modeling_flax_utils.py:542]: Add to do this gimmick since the way weight sharing in `FlaxBert` was done doesn't work for Electra and other models since they don't explictly separate the bias from the kernel (as is the case in Bert)

File: tests/test_modeling_flax_common.py [hunk 5]
```diff
@@ -111,7 +111,7 @@ def test_equivalence_pt_to_flax(self):
                 fx_outputs = fx_model(**prepared_inputs_dict)
                 self.assertEqual(len(fx_outputs), len(pt_outputs), "Output lengths differ between Flax and PyTorch")
                 for fx_output, pt_output in zip(fx_outputs, pt_outputs):
-                    self.assert_almost_equals(fx_output, pt_output.numpy(), 1e-3)
+                    self.assert_almost_equals(fx_output, pt_output.numpy(), 5e-3)
```
Comment (CoderPat) [inline tests/test_modeling_flax_common.py:114]: I changed the threshold because one of the tests was barely failing. I saw that one of the other assertions had this value so I changed them all to that value for consistency. Is that a problem?

File: tests/test_modeling_flax_electra.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    )
+
+    def setUp(self):
+        self.model_tester = FlaxElectraModelTester(self)
+
+    @slow
+    def test_model_from_pretrained(self):
+        for model_class_name in self.all_model_classes:
+            if model_class_name == FlaxElectraForMaskedLM:
+                model = model_class_name.from_pretrained("google/electra-small-generator", from_pt=True)
```
Comment (CoderPat) [inline tests/test_modeling_flax_electra.py:129]: This should not be from_pt, but I had no way to test it without it

Comment (patrickvonplaten) [reply tests/test_modeling_flax_electra.py:129]: I'll upload the Flax weights

File: src/transformers/modeling_flax_utils.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_flax_utils.py:435]: Could we maybe leave it out for now and only add it once needed? :-) I think it's not a good idea to have unused/dead code in the repo

File: tests/test_modeling_flax_common.py [hunk 7]
```diff
@@ -152,7 +152,7 @@ def test_equivalence_flax_to_pt(self):
...
```
Comment (patrickvonplaten) [inline tests/test_modeling_flax_common.py:155]: I guess it's fine to have `5e-3` if ELECTRA needs it. However, the smaller the error tolerance, the better so if tests pass with `1e-3` than let's leave them at `1e-3` :-)

File: src/transformers/modeling_flax_utils.py [hunk 8]
```diff
@@ -55,6 +58,10 @@
 }
 
 
+def identity(x, **kwargs):
```
Comment (patrickvonplaten) [inline src/transformers/modeling_flax_utils.py:61]: could we maybe also remove this for now? :-)

File: docs/source/index.rst [hunk 9]
```diff
@@ -292,7 +292,7 @@ TensorFlow and/or Flax.
 +-----------------------------+----------------+----------------+-----------------+--------------------+--------------+
 |         DistilBERT          |       ✅       |       ✅       |       ✅        |         ✅         |      ❌      |
 +-----------------------------+----------------+----------------+-----------------+--------------------+--------------+
-|           ELECTRA           |       ✅       |       ✅       |       ✅        |         ✅         |      ❌      |
+|           ELECTRA           |       ✅       |       ✅       |       ✅        |         ✅         |      ✅      |
```
Comment (patrickvonplaten) [inline docs/source/index.rst:295]: Yaaaay :-)

File: src/transformers/models/electra/modeling_flax_electra.py [hunk 10]
```diff
@@ -0,0 +1,689 @@
...
+        # Sum all embeddings
+        hidden_states = inputs_embeds + token_type_embeddings + position_embeds
+
+        # Layer Norm
+        hidden_states = self.LayerNorm(hidden_states)
+        hidden_states = self.dropout(hidden_states, deterministic=deterministic)
+        return hidden_states
+
+
+class FlaxElectraSelfAttention(nn.Module):
```
Comment (patrickvonplaten) [inline src/transformers/models/electra/modeling_flax_electra.py:139]: To ensure that Electra and BERT are in sync, the PyTorch modeling file has a bunch of `#Copied from statements` as shown here. Could we add those here as well?

File: src/transformers/models/electra/modeling_flax_electra.py [hunk 11]
```diff
@@ -0,0 +1,689 @@
...
+
+    def __call__(
+        self, input_ids, attention_mask=None, token_type_ids=None, position_ids=None, deterministic: bool = True
+    ):
+        hidden_states = self.electra(
+            input_ids, attention_mask, token_type_ids, position_ids, deterministic=deterministic
+        )[0]
+        prediction_scores = self.generator_predictions(hidden_states)
+
+        if self.config.tie_word_embeddings:
```
Comment (patrickvonplaten) [inline src/transformers/models/electra/modeling_flax_electra.py:478]: Actually, Electra doesn't tie the word embeddings at all, so I don't think we need the logic here :-)

To give a bit more detail here:

In PyTorch, we call the function [_tie_or_clone_weights()](https://github.com/huggingface/transformers/blob/04ab2ca639ee6fd1002ce0a498d892245f0c9093/src/transformers/modeling_utils.py#L581) on every model class that:
- has `self.config.tie_word_embeddings` set to True (this is the case for ELECTRA)
- **and** that overwrites the function `get_output_embeddings()` to return output embeddings != None (this is **not** the case for ELECTRA)

=> no weights are tied

File: src/transformers/modeling_flax_utils.py (hunk 4 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_flax_utils.py:542]: This looks like a cool class! However since ELECTRA doesn't tie word embeddings, I don't think we need it at the moment :-) Could we maybe leave it out for now and remove the tie embedding logic in ELECTRA?

**Review by patrickvonplaten — COMMENTED** [review]
Hey @CoderPat,

Great job, coding up FlaxElectra! This looks very clean & nice :-) 
I think ELECTRA doesn't tie the input & output embeddings (@LysandreJik, @stefan-it correct me if I'm wrong please), so I think we can remove this logic here.

After that I think we only need to add a couple of nits & are then good to go :-)

File: tests/test_modeling_flax_common.py (hunk 7 above)
Comment (patrickvonplaten) [reply tests/test_modeling_flax_common.py:155]: @LysandreJik - If I remember correctly, ELECTRA had very high logit values which could explain the required higher tolerance here - is that correct?

File: src/transformers/modeling_flax_utils.py (hunk 3 above)
Comment (CoderPat) [reply src/transformers/modeling_flax_utils.py:435]: Maybe I was unclear, this is in fact used for now (see [this](https://github.com/huggingface/transformers/blob/f41e4db62662913e36a069910a97770da62b159f/src/transformers/models/electra/modeling_flax_electra.py#L564)). What I meant is that maybe I could have done it without introducing another class, but this way its "scales" more easily to new models

File: tests/test_modeling_flax_common.py (hunk 7 above)
Comment (CoderPat) [reply tests/test_modeling_flax_common.py:155]: Okok, I'll check which ones actually need to be changed. Btw is there a reason for the explained different between Jax and Pytorch? if I understand correctly the weights are exactly the same

File: src/transformers/modeling_flax_utils.py (hunk 8 above)
Comment (CoderPat) [reply src/transformers/modeling_flax_utils.py:61]: this is needed for the `SequenceSummary`, but maybe there is a jax/flax version that I missed (needs to have the kwargs thing)

File: src/transformers/models/electra/modeling_flax_electra.py (hunk 10 above)
Comment (CoderPat) [reply src/transformers/models/electra/modeling_flax_electra.py:139]: Ye ye sure, ill copy them

File: src/transformers/models/electra/modeling_flax_electra.py (hunk 11 above)
Comment (CoderPat) [reply src/transformers/models/electra/modeling_flax_electra.py:478]: Hmm this is in fact weird. I originally didn't include the tied embeddings but I wasn't passing the test without it. I'll try again removing this

File: src/transformers/modeling_flax_utils.py (hunk 3 above)
Comment (sgugger) [reply src/transformers/modeling_flax_utils.py:435]: The PyTorch `SequenceSummary` was a mistake and we will remove it at some point. It goes against our philosophy of having one independent file per model and no abstraction for modeling classes.

So I would not add this here and just define the needed class for Electra in the electra modeling file, ignore the `config.summary_type` argument and just implementing the "first" type. It should also be named with the `FlaxElectra` prefix.

**Review by sgugger — APPROVED** [review]
Thanks for the PR! 
There is one point that reproduces a mistake made in PyTorch that should be removed from the PR. Please take note of our [philosophy](https://huggingface.co/transformers/philosophy.html) which states that every model should have independent modeling files (for easy customization) and no abstract objects. The `SequenceSummary` goes against it.

File: tests/test_modeling_flax_common.py (hunk 7 above)
Comment (patrickvonplaten) [reply tests/test_modeling_flax_common.py:155]: It's rounding errors that can add up

File: src/transformers/modeling_flax_utils.py (hunk 3 above)
Comment (CoderPat) [reply src/transformers/modeling_flax_utils.py:435]: Makes alot of sense, I will move the class to the proper file and rename it (however I'll keep the variable name fore consistency between checkpoints)

File: src/transformers/models/electra/modeling_flax_electra.py (hunk 11 above)
Comment (CoderPat) [reply src/transformers/models/electra/modeling_flax_electra.py:478]: So I just double checked and it seems that weight tying does happen in electra (see [here](https://github.com/huggingface/transformers/blob/bc2571e61c985ec82819cf01ad038342771c94d0/src/transformers/models/electra/modeling_electra.py#L1087)) and I cannot pass the tests without. However, following @sgugger advice, I moved the `TiedDense` inside the flax_electra_module (if you prefered to keep it somewhere reusable let me know)

File: src/transformers/models/electra/modeling_flax_electra.py (hunk 10 above)
Comment (CoderPat) [reply src/transformers/models/electra/modeling_flax_electra.py:139]: (also just realised their true purpose, to ensure consistency between copies, that's super smart!)

**Comment by CoderPat** [comment]
I think I address most comments (the only thing missing is removing the `from_pt` when the flax checkpoint gets upload)! Thanks for the swift feedback

File: tests/test_modeling_flax_electra.py (hunk 6 above)
Comment (CoderPat) [reply tests/test_modeling_flax_electra.py:129]: Thanks! Let me know when you do so I can change the tests

File: src/transformers/models/electra/modeling_flax_electra.py (hunk 10 above)
Comment (patrickvonplaten) [reply src/transformers/models/electra/modeling_flax_electra.py:142]: kudos go to @sgugger haha

File: src/transformers/models/electra/modeling_flax_electra.py (hunk 11 above)
Comment (patrickvonplaten) [reply src/transformers/models/electra/modeling_flax_electra.py:508]: You are totally right - thanks for double checking! I missed the `get_output_embeddings()` function

File: tests/test_modeling_flax_electra.py (hunk 6 above)
Comment (patrickvonplaten) [reply tests/test_modeling_flax_electra.py:129]: `"google/electra-small-generator"` is uploaded

File: tests/test_modeling_flax_common.py (hunk 7 above)
Comment (LysandreJik) [reply tests/test_modeling_flax_common.py]: It has logits in the range of +-1e[0-1], so 5e-3 is definitely reasonable!

File: docs/source/index.rst (hunk 9 above)
Comment (LysandreJik) [reply docs/source/index.rst:295]: :heart_eyes:

**Review by LysandreJik — APPROVED** [review]
Love it, LGTM!

File: src/transformers/models/electra/modeling_flax_electra.py [hunk 12]
```diff
@@ -0,0 +1,1146 @@
...
+            Hidden-states of the model at the output of each layer plus the initial embedding outputs.
+        attentions (:obj:`tuple(jax_xla.DeviceArray)`, `optional`, returned when ``output_attentions=True`` is passed or when ``config.output_attentions=True``):
+            Tuple of :obj:`jax_xla.DeviceArray` (one for each layer) of shape :obj:`(batch_size, num_heads,
+            sequence_length, sequence_length)`.
+
+            Attentions weights after the attention softmax, used to compute the weighted average in the self-attention
+            heads.
+    """
+
+    logits: jax_xla.DeviceArray = None
```
Comment (CoderPat) [inline src/transformers/models/electra/modeling_flax_electra.py:78]: I renamed this from `prediction_logits` to `logits` to be consistent with the Pytorch Electra

Comment (patrickvonplaten) [reply src/transformers/models/electra/modeling_flax_electra.py:78]: great!

File: src/transformers/models/electra/modeling_flax_electra.py [hunk 13]
```diff
@@ -0,0 +1,1146 @@
...
+        position_ids (:obj:`numpy.ndarray` of shape :obj:`({0})`, `optional`):
+            Indices of positions of each input sequence tokens in the position embeddings. Selected in the range ``[0,
+            config.max_position_embeddings - 1]``.
+        return_dict (:obj:`bool`, `optional`):
+            Whether or not to return a :class:`~transformers.file_utils.ModelOutput` instead of a plain tuple.
+
+"""
+
+
+class FlaxElectraEmbeddings(nn.Module):
```
Comment (patrickvonplaten) [inline src/transformers/models/electra/modeling_flax_electra.py:141]: This looks identical to `FlaxBertEmbeddings` to me -> could we add a `# Copied from ...` statement? :-)

Comment (CoderPat) [reply src/transformers/models/electra/modeling_flax_electra.py:141]: It's not actually the same, it uses `self.config.vocab_size` rather than `self.config.hidden_size`

Comment (patrickvonplaten) [reply src/transformers/models/electra/modeling_flax_electra.py:141]: Yes! Just saw it :-) - added a # copied from to the forward, that should be good enough

**Review by patrickvonplaten — APPROVED** [review]
Great job @CoderPat
